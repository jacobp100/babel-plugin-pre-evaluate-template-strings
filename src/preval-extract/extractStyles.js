/* @flow */

import fs from 'fs';
import path from 'path';
import mkdirp from 'mkdirp';

import type { BabelTypes, NodePath } from '../types';

import { getCachedModule } from '../lib/moduleSystem';
import { relativeToCwd, makeAbsolute } from './utils';

const withPreamble = data =>
  `/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */\n${data}`;

/**
 * Get output filename with directory structure from source preserved inside
 * custom outDir.
 */
function getOutputFilename(
  relativeFilename: string,
  absOutDir: string
): string {
  const basename = /(.+)\..+$/.exec(path.basename(relativeFilename))[1];
  const relativeOutputFilename = path.join(
    path.dirname(relativeFilename),
    `${basename}.css`
  );
  return path.join(absOutDir, relativeOutputFilename);
}

let stylesCache = {};

function hasCachedStyles(filename: string, styles: string) {
  return stylesCache[filename] && stylesCache[filename] === styles;
}

function createCssFromCache() {
  return withPreamble(
    Object.keys(stylesCache).reduce(
      (acc, file) => `${acc}\n${stylesCache[file]}`,
      ''
    )
  );
}

function addRequireForCss(
  types: BabelTypes,
  program: NodePath<*>,
  filename: string
) {
  program.node.body.unshift(
    types.expressionStatement(
      types.callExpression(types.identifier('require'), [
        types.stringLiteral(filename),
      ])
    )
  );
}

/**
 * Write styles to file and create directory if needed.
 */
function outputStylesToFile(
  filename: string,
  styles: string,
  throwImmediately: boolean = false
) {
  try {
    fs.writeFileSync(filename, styles);
  } catch (error) {
    if (
      !throwImmediately &&
      (error.code === 'ENOENT' || /ENOENT/.test(error.message))
    ) {
      mkdirp.sync(path.dirname(filename));
      outputStylesToFile(filename, styles, true);
    } else {
      throw error;
    }
  }
}

export default function extractStyles(
  types: BabelTypes,
  program: NodePath<*>,
  currentFilename: string,
  options: {
    single?: boolean,
    filename?: string,
    outDir?: string,
    cache?: boolean,
    extract?: boolean,
  }
) {
  // Normalize current filename path.
  const relativeCurrentFilename = relativeToCwd(currentFilename);
  const absCurrentFilename = makeAbsolute(currentFilename);

  const {
    single = false,
    cache = true,
    extract = true,
    outDir = '.linaria-cache',
    filename: basename = 'styles.css',
  } = options;

  const absOutDir = path.isAbsolute(outDir)
    ? outDir
    : path.join(process.cwd(), outDir);

  // If single === true, we compute filename from outDir and filename options,
  // since there will be only one file, otherwise we need to reconstruct directory
  // structure inside outDir. In that case filename option is discard.
  const filename = single
    ? path.join(absOutDir, basename)
    : getOutputFilename(relativeCurrentFilename, absOutDir);
  const importPath = `./${path.relative(
    path.dirname(absCurrentFilename),
    filename
  )}`;

  if (!extract) {
    return;
  }

  const sheet = getCachedModule(require.resolve('../../sheet.js'));
  const data = sheet ? sheet.exports.default.dump() : '';

  if (!data.length) {
    return;
  }

  if (cache) {
    if (single) {
      // If single === true, we cannot rely on filename since it will
      // always be the same, so we need to use absCurrentFilename.
      if (hasCachedStyles(absCurrentFilename, data)) {
        return;
      }
      stylesCache[absCurrentFilename] = data;
    } else {
      if (hasCachedStyles(filename, data)) {
        addRequireForCss(types, program, importPath);
        return;
      }
      stylesCache[filename] = data;
    }
  }

  if (single) {
    outputStylesToFile(
      filename,
      cache ? createCssFromCache() : withPreamble(data)
    );
  } else {
    outputStylesToFile(filename, withPreamble(data));
    addRequireForCss(types, program, importPath);
  }
}

export function clearCache() {
  stylesCache = {};
}
