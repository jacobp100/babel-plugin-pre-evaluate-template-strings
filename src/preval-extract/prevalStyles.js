/* @flow */

import { resolve } from 'path';
import generate from 'babel-generator';
import shortHash from 'short-hash';

import type {
  BabelCore,
  State,
  NodePath,
  BabelTaggedTemplateExpression,
  BabelIdentifier,
  BabelCallExpression,
  RequirementSource,
} from '../types';

import getReplacement from './getReplacement';
import {
  instantiateModule,
  clearLocalModulesFromCache,
} from '../lib/moduleSystem';

function getMinifiedClassName(className: string) {
  return `ln${shortHash(className)}`;
}

/**
 * const header = css`
 *   color: ${header.color};
 * `;
 *
 * const header = preval`
 *   module.exports = css.named('header_slug')`
 *     color: ${header.color}
 *   `;
 * `;
 */

export default function(
  babel: BabelCore,
  title: string,
  path: NodePath<
    BabelTaggedTemplateExpression<BabelIdentifier | BabelCallExpression>
  >,
  state: State,
  requirements: RequirementSource[]
) {
  const { name } = path.scope.generateUidIdentifier(title);
  const source = path.getSource() || generate(path.node).code;

  const replacement = getReplacement([
    ...requirements,
    {
      code: `module.exports = ${source
        .replace(/css(?!\.named)/g, `css.named('${name}', '${state.filename}')`)
        .replace(
          /css\.named\(([^,]+)\)/,
          (input, customName) => `css.named(${customName}, '${state.filename}')`
        )}`,
      loc: path.node.loc.start,
    },
  ]);

  clearLocalModulesFromCache();
  const { exports } = instantiateModule(replacement, resolve(state.filename));

  return exports;
}
