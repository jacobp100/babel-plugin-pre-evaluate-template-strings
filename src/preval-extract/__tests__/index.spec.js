import * as babel from 'babel-core';
import prevalExtractPlugin, {
  cssTaggedTemplateRequirementsVisitor,
  externalRequirementsVisitor,
} from '../index';
import extractStyles from '../extractStyles';
import prevalStyles from '../prevalStyles';

jest.mock('../extractStyles');
jest.mock('../prevalStyles');

describe('preval-extract/index', () => {
  describe('Program#enter visitor', () => {
    it('should set necessary flags in state', () => {
      const { visitor } = prevalExtractPlugin(babel);

      const state = {
        file: {
          opts: {
            filename: 'test.js',
          },
        },
      };

      visitor.Program.enter(
        {
          container: {
            tokens: [],
          },
        },
        state
      );

      expect(state).toEqual({
        skipFile: false,
        foundLinariaTaggedLiterals: false,
        filename: 'test.js',
        file: {
          opts: {
            filename: 'test.js',
          },
        },
      });
    });

    it('should set skipFile flag if linaria-preval comment is found', () => {
      const { visitor } = prevalExtractPlugin(babel);

      const state = {
        file: {
          opts: {
            filename: 'test.js',
          },
        },
      };

      visitor.Program.enter(
        {
          container: {
            tokens: [
              {
                type: 'CommentBlock',
                value: ' linaria-preval ',
              },
            ],
          },
        },
        state
      );

      expect(state.skipFile).toBeTruthy();
    });
  });

  describe('Program#exit visitor', () => {
    beforeEach(() => {
      extractStyles.mockClear();
    });

    it('should call extractStyles if linaria tagged template literal was found', () => {
      const { visitor } = prevalExtractPlugin(babel);

      visitor.Program.exit(
        {
          container: {
            tokens: [],
          },
        },
        {
          skipFile: false,
          foundLinariaTaggedLiterals: true,
        }
      );

      expect(extractStyles).toHaveBeenCalled();
    });

    it('should not call extractStyles there were not styles extracted', () => {
      const { visitor } = prevalExtractPlugin(babel);

      visitor.Program.exit(
        {
          container: {
            tokens: [],
          },
        },
        {
          skipFile: false,
          foundLinariaTaggedLiterals: false,
        }
      );

      expect(extractStyles).not.toHaveBeenCalled();
    });

    it('should not do anything is skipFile flag is true', () => {
      const { visitor } = prevalExtractPlugin(babel);

      visitor.Program.exit(
        {
          container: {
            tokens: [],
          },
        },
        {
          skipFile: true,
          get foundLinariaTaggedLiterals() {
            throw new Error(
              'foundLinariaTaggedLiterals should not have been called'
            );
          },
        }
      );

      expect(extractStyles).not.toHaveBeenCalled();
    });
  });

  describe('TaggedTemplateExpression', () => {
    beforeEach(() => {
      prevalStyles.mockClear();
    });

    it('should not do anything if skipFile flag is set to true', () => {
      const { visitor } = prevalExtractPlugin(babel);

      visitor.TaggedTemplateExpression(null, {
        skipFile: true,
        foundLinariaTaggedLiterals: false,
      });

      expect(prevalStyles).not.toHaveBeenCalled();
    });

    it('should not process template if checks do not pass', () => {
      const { visitor } = prevalExtractPlugin(babel);

      visitor.TaggedTemplateExpression(
        {
          node: {
            type: 'TaggedTemplateExpression',
            tag: {
              type: 'Identifier',
              name: 'styles',
            },
          },
        },
        {
          skipFile: false,
          foundLinariaTaggedLiterals: false,
        }
      );

      expect(prevalStyles).not.toHaveBeenCalled();

      expect(() => {
        visitor.TaggedTemplateExpression(
          {
            node: {
              type: 'TaggedTemplateExpression',
              tag: {
                type: 'Identifier',
                name: 'css',
              },
            },
            findParent: () => null,
            buildCodeFrameError: text => new Error(text),
          },
          {
            skipFile: false,
            foundLinariaTaggedLiterals: false,
          }
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('should prevalStyles with variable assignment', () => {
      const { visitor } = prevalExtractPlugin(babel);

      const parent = {
        type: 'VariableDeclarator',
        node: {
          type: 'VariableDeclarator',
          id: {
            name: 'test',
          },
        },
      };

      visitor.TaggedTemplateExpression(
        {
          node: {
            type: 'TaggedTemplateExpression',
            tag: {
              type: 'Identifier',
              name: 'css',
            },
          },
          traverse: () => {},
          findParent: check => (check(parent) ? parent : null),
          replaceWith: () => {},
          addComment: () => {},
        },
        {
          skipFile: false,
          foundLinariaTaggedLiterals: false,
        }
      );

      expect(prevalStyles).toHaveBeenCalled();
      expect(prevalStyles.mock.calls[0][1]).toBe('test');
    });

    it('should prevalStyles with object property with identifier', () => {
      const { visitor } = prevalExtractPlugin(babel);

      const parent = {
        type: 'ObjectProperty',
        node: {
          type: 'ObjectProperty',
          key: {
            type: 'Identifier',
            name: 'foo',
          },
        },
      };

      visitor.TaggedTemplateExpression(
        {
          node: {
            type: 'TaggedTemplateExpression',
            tag: {
              type: 'Identifier',
              name: 'css',
            },
          },
          traverse: () => {},
          findParent: check => (check(parent) ? parent : null),
          replaceWith: () => {},
          addComment: () => {},
        },
        {
          skipFile: false,
          foundLinariaTaggedLiterals: false,
        }
      );

      expect(prevalStyles).toHaveBeenCalled();
      expect(prevalStyles.mock.calls[0][1]).toBe('foo');
    });

    it('should prevalStyles with object property with string literal', () => {
      const { visitor } = prevalExtractPlugin(babel);

      const parent = {
        type: 'ObjectProperty',
        node: {
          type: 'ObjectProperty',
          key: {
            type: 'StringLiteral',
            value: 'foo',
          },
        },
      };

      visitor.TaggedTemplateExpression(
        {
          node: {
            type: 'TaggedTemplateExpression',
            tag: {
              type: 'Identifier',
              name: 'css',
            },
          },
          traverse: () => {},
          findParent: check => (check(parent) ? parent : null),
          replaceWith: () => {},
          addComment: () => {},
        },
        {
          skipFile: false,
          foundLinariaTaggedLiterals: false,
        }
      );

      expect(prevalStyles).toHaveBeenCalled();
      expect(prevalStyles.mock.calls[0][1]).toBe('foo');
    });

    it('should prevalStyles with object property with numeric literal', () => {
      const { visitor } = prevalExtractPlugin(babel);

      const parent = {
        type: 'ObjectProperty',
        node: {
          type: 'ObjectProperty',
          key: {
            type: 'NumericLiteral',
            value: 42,
          },
        },
      };

      visitor.TaggedTemplateExpression(
        {
          node: {
            type: 'TaggedTemplateExpression',
            tag: {
              type: 'Identifier',
              name: 'css',
            },
          },
          traverse: () => {},
          findParent: check => (check(parent) ? parent : null),
          replaceWith: () => {},
          addComment: () => {},
        },
        {
          skipFile: false,
          foundLinariaTaggedLiterals: false,
        }
      );

      expect(prevalStyles).toHaveBeenCalled();
      expect(prevalStyles.mock.calls[0][1]).toBe(42);
    });

    it('should prevalStyles with JSX opening element', () => {
      const { visitor } = prevalExtractPlugin(babel);

      const parent = {
        type: 'JSXOpeningElement',
        node: {
          type: 'JSXOpeningElement',
          name: {
            type: 'JSXIdentifier',
            name: 'article',
          },
        },
      };

      visitor.TaggedTemplateExpression(
        {
          node: {
            type: 'TaggedTemplateExpression',
            tag: {
              type: 'Identifier',
              name: 'css',
            },
          },
          traverse: () => {},
          findParent: check => (check(parent) ? parent : null),
          replaceWith: () => {},
          addComment: () => {},
        },
        {
          skipFile: false,
          foundLinariaTaggedLiterals: false,
        }
      );

      expect(prevalStyles).toHaveBeenCalled();
      expect(prevalStyles.mock.calls[0][1]).toBe('article');
    });
  });

  [
    {
      id: 'cssTaggedTemplateRequirementsVisitor',
      visitor: cssTaggedTemplateRequirementsVisitor,
    },
    {
      id: 'externalRequirementsVisitor',
      visitor: externalRequirementsVisitor,
    },
  ].forEach(({ id, visitor }) => {
    describe(id, () => {
      it('should not do anything if path is not referenced or is excluded', () => {
        const self = {};
        visitor.Identifier.call(self, {
          isReferenced: () => false,
        });

        visitor.Identifier.call(self, {
          isReferenced: () => true,
          node: {
            name: 'test',
          },
          scope: {
            getBinding: () => ({ kind: 'param' }),
          },
        });

        expect(self).toEqual({});
      });

      it('should not add source to requirements if it is null', () => {
        const self = { requirements: [] };
        visitor.Identifier.call(self, {
          isReferenced: () => true,
          node: {
            name: 'test',
          },
          scope: {
            getBinding: () => ({
              kind: 'expression',
              path: {
                getSource: () => null,
                node: {
                  loc: null,
                },
              },
            }),
          },
        });

        expect(self).toEqual({ requirements: [] });
      });

      it('should not add source to requirements if it was already added', () => {
        const requirements = [{ code: 'source', loc: { line: 0, column: 0 } }];
        const self = {
          requirements: [...requirements],
        };
        visitor.Identifier.call(self, {
          isReferenced: () => true,
          node: {
            name: 'test',
          },
          scope: {
            getBinding: () => ({
              kind: 'expression',
              path: {
                getSource: () => 'source',
                node: {
                  loc: { line: 0, column: 0 },
                },
              },
            }),
          },
        });

        expect(self).toEqual({ requirements });
      });

      it('should add source to requirements', () => {
        const self = { requirements: [], addBeforeIndex: 0 };
        visitor.Identifier.call(self, {
          isReferenced: () => true,
          node: {
            name: 'test',
          },
          scope: {
            getBinding: () => ({
              kind: 'expression',
              path: {
                getSource: () => 'source',
                isImportDefaultSpecifier: () => true,
                node: {
                  loc: { start: { line: 0, column: 0 } },
                },
              },
            }),
          },
        });

        expect(self).toEqual({
          addBeforeIndex: 0,
          requirements: [{ code: 'source', loc: { line: 0, column: 0 } }],
        });
      });

      it('should add source to requirements and traverse with external visitor', () => {
        const traverse = jest.fn();
        const self = { requirements: [], addBeforeIndex: 0 };
        visitor.Identifier.call(self, {
          isReferenced: () => true,
          node: {
            name: 'test',
          },
          scope: {
            getBinding: () => ({
              kind: 'expression',
              path: {
                getSource: () => 'source',
                isImportDefaultSpecifier: () => false,
                isImportSpecifier: () => false,
                node: {
                  loc: { start: { line: 0, column: 0 } },
                },
                traverse,
              },
            }),
          },
        });

        expect(self).toEqual({
          addBeforeIndex: 0,
          requirements: [{ code: 'source', loc: { line: 0, column: 0 } }],
        });
        expect(traverse).toHaveBeenCalled();
      });
    });
  });
});
