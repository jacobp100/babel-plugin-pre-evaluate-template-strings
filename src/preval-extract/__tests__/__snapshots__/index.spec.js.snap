// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preval-extract/index TaggedTemplateExpression should not process template if checks do not pass 1`] = `
"Couldn't determine the class name for CSS template literal. Ensure that it's either:
- Assigned to a variable
- Is an object property
- Is a prop in a JSX element
"
`;
